arborist/role.go

```diff
@@ -0,0 +1,222 @@
+func (role *Role) permit(permission *Permission) {
+	role.Permissions[permission] = struct{}{}
+}
+
+// hasTags checks if this role contains all the tags in the given list.
+func (role *Role) hasTags(tags []string) bool {
+	result := true
+	for _, tag := range tags {
+		if _, contains := role.Tags[tag]; !contains {
+			result = false
+			break
+		}
+	}
+	return result
+}
+
+// allSubroles returns a slice of all the subroles accessible in the tree
+// beneath this role.
+func (role *Role) allSubroles() []*Role {
+	var result []*Role
+	var queue roleQueue
+	queue.pushBack(role)
+
+	for queue.nonempty() {
+		next := queue.popFront()
+		result = append(result, next)
+		for role := range next.Subroles {
+			queue.pushBack(role)
+		}
+	}
+
+	return result
+}
+
+// update appends the contents of all the fields in `input_role` onto the
+// existing fields in `role`. This can include overwriting the current name with
+// a new name given in the input role.
+func (role *Role) update(input_role *Role) {
+	if input_role.ID != "" {
+		role.ID = input_role.ID
+	}
+
+	for subrole := range input_role.Subroles {
+		role.Subroles[subrole] = struct{}{}
+	}
+
+	for permission := range input_role.Permissions {
+		role.Permissions[permission] = struct{}{}
+	}
+
+	for tag := range input_role.Tags {
+		role.Tags[tag] = struct{}{}
+	}
+}
+
+// allPermissions collects all the permissions this role implies, accumulating
+// the permissions for every subrole starting from this one.
+func (role *Role) allPermissions() Permissions {
+	var permissions Permissions
+
+	for _, subrole := range role.allSubroles() {
+		for permission := range subrole.Permissions {
+			permissions = append(permissions, permission)
+		}
+	}
+
+	return permissions
+}
+
+// validate checks if the attempted action and constraints are allowed under
+// this role.
+func (role *Role) validate(try_action Action, try_constraints Constraints) authResponse {
+	auth := role.allPermissions().validate(try_action, try_constraints)
+	if auth.Auth {
+		auth.Role_ID = &role.ID
+	}
+	return auth
+}
+
+// toJSON converts a `Role` to a `RoleJSON`.
+func (role *Role) toJSON() RoleJSON {
+	var i uint
+
+	subroles := make([]RoleJSON, len(role.Subroles))
+	i = 0
+	for subrole := range role.Subroles {
+		subroles[i] = subrole.toJSON()
+	}
+
+	permissions := make([]PermissionJSON, len(role.Permissions))
+	i = 0
+	for permission := range role.Permissions {
+		permissions[i] = permission.toJSON()
+	}
+
+	tags := make([]string, len(role.Tags))
+	i = 0
+	for tag := range role.Tags {
+		tags[i] = tag
+	}
+
+	return RoleJSON{
+		ID:          role.ID,
+		Subroles:    subroles,
+		Permissions: permissions,
+		Tags:        tags,
+	}
+}
+
+// RoleJSON represents a `Role` in JSON format. In particular, the subroles,
+// permissions, and tags, which are stored as maps in the role, should should
+// just be arrays in the JSON output. This is *only* used for marshalling roles
+// to and from JSON.
+type RoleJSON struct {
```

Comment:
how do I submit a role to arborist specifying it's parent? should this struct has a parent field? and if it's provided, it should try to find that and append the role under the parent instead of the root role?

Reply:
my thought on this was that to add a role under a parent, you make a `PATCH /role/<parent>` to the parent with this:
```json
{
  "subroles": [
    {
      "id": "the new subrole you want to add",
      "etc": "..."
    }
  ]
}
```
so the engine looks up the parent role using the URL, loads the subrole as if it would go under the root, and then inserts the subrole using `Role.insert`

Reply:
oh ok..